[Writing Tools] Two Writing Tools context menu items show up; menu item shows up even in non-text cases; panel appears in wrong location with one of the items; menu item is in wrong location

rr-codes opened the pull request:
#### bf5cf82f2df5bbd2afd6e7cdc56f6da24b2b21fb
<pre>
[Writing Tools] Two Writing Tools context menu items show up; menu item shows up even in non-text cases; panel appears in wrong location with one of the items; menu item is in wrong location
<a href="https://bugs.webkit.org/show_bug.cgi?id=276241">https://bugs.webkit.org/show_bug.cgi?id=276241</a>
<a href="https://rdar.apple.com/127608508">rdar://127608508</a>

Reviewed by Wenson Hsieh, Aditya Keerthi and Tim Horton.

* Implement the `providesWritingToolsContextMenu` property so that AppKit does not inject their own
Writing Tools context menu item.

* Move the Writing Tools context menu item to be just above the &apos;Spelling &amp; Grammar&apos; item.

* Use `standardWritingToolsMenuItem` to implement the new Writing Tools context menu item UX so that
it includes each tool as a sub-menu item.

* Maintain compatibility in cases where these methods are not currently implemented by keeping the
old implementation around temporarily.

* Source/WebCore/PAL/pal/spi/mac/NSMenuSPI.h:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::populate):
* Source/WebCore/page/writing-tools/WritingToolsTypes.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView writingToolsBehavior]):
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView providesWritingToolsContextMenu]):
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.h:
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.mm:
(WebKit::convertToPlatformRequestedTool):
(WebKit::convertToWebRequestedTool):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::writingToolsBehavior const):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::writingToolsContextMenuItemSelected):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::writingToolsContextMenuItemSelected):
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.h:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(-[WKMenuTarget showWritingTools:]):
(WebKit::WebContextMenuProxyMac::writingToolsContextMenuItemSelected):
(WebKit::WebContextMenuProxyMac::getContextMenuItem):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::writingToolsContextMenuItemSelected):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::canHandleContextMenuWritingTools const):

Canonical link: <a href="https://commits.webkit.org/280705@main">https://commits.webkit.org/280705@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1c775e6c350676cc3a57ba99daf2e3e2f835d0b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57389 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36717 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9864 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61011 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7834 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59517 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44341 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8022 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/46473 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5542 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59419 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/34455 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49563 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27337 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31237 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6874 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6837 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7145 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62690 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1302 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7238 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53733 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1308 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49597 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/53821 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1111 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8561 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/32546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33631 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34716 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33377 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/ContextMenuController.cpp:
```diff
@@ -1341,8 +1342,16 @@ void ContextMenuController::populate()
+#if ENABLE(WRITING_TOOLS)
+        appendItem(*separatorItem(), m_contextMenu.get());
+        ContextMenuItem writingToolsItem(ContextMenuItemType::Action, ContextMenuItemTagWritingTools, contextMenuItemTagWritingTools());
+        appendItem(writingToolsItem, m_contextMenu.get());
```
Indentation is off here.

hortont424 commented on Source/WebCore/page/writing-tools/WritingToolsTypes.h:
```diff
@@ -47,6 +47,10 @@ enum class Action : uint8_t {
 };
 
+enum class RequestedTool : uint8_t {
+    // Opaque type to transitively convert to/from WTRquestedTool.
```
Sp Requested?

pxlcoder replied:
```suggestion
    // Opaque type to transitively convert to/from WTRequestedTool.
```

hortont424 commented on Source/WebCore/page/writing-tools/WritingToolsTypes.h:
```diff
@@ -47,6 +47,10 @@ enum class Action : uint8_t {
     Restart,
 };
 
+enum class RequestedTool : uint8_t {
```
Is this big enough?

pxlcoder replied:
Looks like it is for now, but no guarantee in the future.

pxlcoder replied:
Can we just use a larger type?

rr-codes committed:
[Writing Tools] Two Writing Tools context menu items show up; menu item shows up even in non-text cases; panel appears in wrong location with one of the items; menu item is in wrong location
https://bugs.webkit.org/show_bug.cgi?id=276241
rdar://127608508

Reviewed by Wenson Hsieh, Aditya Keerthi and Tim Horton.

* Implement the `providesWritingToolsContextMenu` property so that AppKit does not inject their own
Writing Tools context menu item.

* Move the Writing Tools context menu item to be just above the 'Spelling & Grammar' item.

* Use `standardWritingToolsMenuItem` to implement the new Writing Tools context menu item UX so that
it includes each tool as a sub-menu item.

* Maintain compatibility in cases where these methods are not currently implemented by keeping the
old implementation around temporarily.

* Source/WebCore/PAL/pal/spi/mac/NSMenuSPI.h:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::populate):
* Source/WebCore/page/writing-tools/WritingToolsTypes.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView writingToolsBehavior]):
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView providesWritingToolsContextMenu]):
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.h:
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.mm:
(WebKit::convertToPlatformRequestedTool):
(WebKit::convertToWebRequestedTool):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::writingToolsBehavior const):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::writingToolsContextMenuItemSelected):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::writingToolsContextMenuItemSelected):
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.h:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(-[WKMenuTarget showWritingTools:]):
(WebKit::WebContextMenuProxyMac::writingToolsContextMenuItemSelected):
(WebKit::WebContextMenuProxyMac::getContextMenuItem):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::writingToolsContextMenuItemSelected):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::canHandleContextMenuWritingTools const):

Canonical link: https://commits.webkit.org/280705@main
